substrate/frame/staking/src/pallet/impls.rs

```diff
@@ -1132,6 +1167,49 @@ impl<T: Config> Pallet<T> {
 	) -> Exposure<T::AccountId, BalanceOf<T>> {
 		EraInfo::<T>::get_full_exposure(era, account)
 	}
+
+	/// Whether the passed reward destination is restricted for the given account.
+	///
+	/// virtual stakers are not allowed to compound their rewards as this pallet does not manage
+	/// locks for them. For external pallets that manage the virtual bond, it is their
+	/// responsibility to distribute the reward and re-bond them.
+	///
+	/// Conservatively, we expect them to always set the reward destination to a non stash account.
+	pub(crate) fn restrict_reward_destination(
```

Comment:
not nit: About the usage of this function - this is currently used in `Ledger::set_payee` and `Ledger::bond`, please correct me if I missed anything.

Call instances for this function seem to be as follows:
- for `set_payee`, it's called in the `set_payee` and `update_payee` extrinsics
- for `bond`, it's called in the `bond` extrinsic and in the `virtual_bond` implementation

It seems to me that we could move this check to the callers. Functionally it would be the same, but it's probably the responsibility of the callers to ensure a correct reward destination rather than the ledger, which doesn't even know why the destination is restricted.